### PR TITLE
Upgrade: watch plan secrets for pre-drain/post-drain hooks

### PR DESCRIPTION
**Pull request by bk201**
**Problem:**

Rancher changed the annotated resource from CAPI machine to machine
plane secrets. 

**Solution:**
Adapt this change. Watch plan secrets like:
```
╰─$ k get secret -n fleet-local --field-selector type=rke.cattle.io/machine-plan
NAME                               TYPE                         DATA   AGE
custom-3e7f6c3bb092-machine-plan   rke.cattle.io/machine-plan   11     8h
custom-b4f71fe2e772-machine-plan   rke.cattle.io/machine-plan   11     8h
custom-d1b67c54e39d-machine-plan   rke.cattle.io/machine-plan   11     8h
custom-fe5fae28238d-machine-plan   rke.cattle.io/machine-plan   11     8h
```

Instead of watching CAPI machines:
```
╰─$ k get machines -n fleet-local
NAME                  CLUSTER   NODENAME   PROVIDERID     PHASE     AGE   VERSION
custom-3e7f6c3bb092   local     node4      rke2://node4   Running   8h
custom-b4f71fe2e772   local     node2      rke2://node2   Running   8h
custom-d1b67c54e39d   local     node3      rke2://node3   Running   8h
custom-fe5fae28238d   local     node1      rke2://node1   Running   8h
```

**Related Issue:**

https://github.com/harvester/harvester/issues/1935

**Test plan:**

### CODE REVIEW

File: pkg/controller/master/upgrade/job_controller.go [hunk 1]
```diff
@@ -159,23 +160,30 @@ func (h *jobHandler) syncNodeJob(job *batchv1.Job) (*batchv1.Job, error) {
...
-	if err != nil {
-		return job, err
+	// find machine plan secret
+	secrets, err := h.secretClient.List(rancherPlanSecretNamespace, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", rancherPlanSecretMachineLabel, machineName),
+		FieldSelector: fmt.Sprintf("type=%s", rancherPlanSecretType),
+	})
+
+	if err != nil || len(secrets.Items) != 1 {
+		return job, nil
```
Comment (weihanglo) [inline pkg/controller/master/upgrade/job_controller.go]: Is this intentional not to return `err`?
```suggestion
	if err != nil || len(secrets.Items) != 1 {
		return job, err
```

File: pkg/controller/master/upgrade/node_controller.go [hunk 2]
```diff
@@ -64,11 +63,19 @@ func (h *nodeHandler) OnChanged(key string, node *corev1.Node) (*corev1.Node, er
...
-		return node, err
+
+	// find machine plan secret
+	secrets, err := h.secretClient.List(rancherPlanSecretNamespace, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", rancherPlanSecretMachineLabel, machineName),
+		FieldSelector: fmt.Sprintf("type=%s", rancherPlanSecretType),
+	})
+
+	if err != nil || len(secrets.Items) != 1 {
+		return node, nil
```
Comment (weihanglo) [inline pkg/controller/master/upgrade/node_controller.go]: Ditto. Return `err`?

```suggestion
	if err != nil || len(secrets.Items) != 1 {
		return node, err
```

File: pkg/controller/master/upgrade/job_controller.go (hunk 1 above)
Comment (FrankYang0529) [reply pkg/controller/master/upgrade/job_controller.go]: If the length of `secrets.Items` is more than 1, do we need to return err? If yes, we may need to separate this block.

Comment (bk201) [reply pkg/controller/master/upgrade/job_controller.go]: No, it's a programming error.
An error or len(secrets) != 1 should both return errrors. Will fix it.

Comment (bk201) [reply pkg/controller/master/upgrade/job_controller.go]: @FrankYang0529 yes, need to create an error when len(items) != 1.

Comment (bk201) [reply pkg/controller/master/upgrade/job_controller.go]: Fixed, please see https://github.com/harvester/harvester/compare/577b8d4214bf3904cabec39472a077bdeedfc97d..324d08231c8fa1e8061ca9e286899147deb9c5c5

**Review by FrankYang0529 — APPROVED** [review]
Test with the following steps and it works. LGTM.

- Create a master-head cluster.
- Run `make build-iso` and `make default`.
- Replace the image in the `harvester-system/harvester` deployment.
- Tag current-version for harvester-upgrade image:
  - Get the current version:
  ```
  $ kubectl get settings server-version
  NAME             VALUE
  server-version   324d0823
  ```
  - **On each node**
  ```
  $ docker tag rancher/harvester-upgrade:master-head rancher/harvester-upgrade:324d0823
  ```

- Create the version:
```
$ cat version.yaml
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: 1.0.2
  namespace: harvester-system
spec:
  isoChecksum: xxx                              <-- the build ISO sha512 checksum
  isoURL: http://10.10.0.1/harvester/harvester.iso      <-- the build ISO
  minUpgradableVersion: 0.3.0
  releaseDate: "20211231"
  tags:
  - dev                 <--- `dev` tag is important
  - test

 $ kubectl create -f version.yaml
```

- Create an upgrade:
```
$ cat upgrade.yaml
apiVersion: harvesterhci.io/v1beta1
kind: Upgrade
metadata:
  generateName: hvst-upgrade-
  namespace: harvester-system
spec:
  version: "1.0.2"

$ kubectl create -f upgrade.yaml
```

**Review by weihanglo — APPROVED** [review]
LGTM. Though sometimes curl corrupted but it is out of scope of this patch IMO.

**Pull request merged**